StoDevX/AAO-React-Native
# 🧪  internal developer feature flagging

Links
- Part of #6106 
- Eventually will impact #6936

[Best viewed without whitespace
](https://github.com/StoDevX/AAO-React-Native/pull/7087/files?diff=split&w=1)

Changes
- Allows for us to define hooks for stating that an internal setting should be toggled in the app.
- Introduces `@frogpond/app-config` module
- Adds a section list of boolean toggles to `Settings > Developer > Feature Flags`
- Output of individual config checks includes a check for `devMode()` enabled
- Adds a `disabled` property to homescreen tiles
- Adds a feature flag for navigating directly to course search without showing the interstitial recents view from the Home Screen
- Adds react-native-testing-library as a dependency

Note

I had this setup with a refactored version of stoPrint but changes were peppered everywhere. This is setup separately from those changes, although the commits are still in the history of this PR. If we accept this method/setup, I'll be sure to squash commit to remove that history.

Intended use 

Change certain sections of the app for more developer exposure and testing to WIP features and layout changes. This is not intended to introduce hidden features or toggles for A/B testing in production.  

Challenges

Hooks are great but they cannot be used everywhere because [hooks have rules](https://legacy.reactjs.org/docs/hooks-rules.html). It will be good to see if this is a useful setup or if it will have to adapt to support areas where hooks cannot be used. One example is the movies tab -- we cannot simply drop this hook into the tab definition to enable/disable it.
 
---

Usage

1. Add types to `modules/app-config/types.ts`

The format is SECTION_KEY where the first underscore will be split at render time for our view to group by, but the entirety of the value (section + key) will be stored.
```tsx
export enum AppConfigType {
	Courses_ShowRecentSearchScreen = 'Courses_ShowRecentSearchScreen',
}
```

2. Add AppConfig definition and feature hook to `modules/app-config/index.tsx`

```tsx
{
	title: 'Show the course search recents screen',
	configKey: AppConfigType.Courses_ShowRecentSearchScreen,
	active: await getFeatureFlag(
		AppConfigType.Courses_ShowRecentSearchScreen,
	),
},

export const useCourseSearchRecentsScreen = (): boolean =>
	useFeature(AppConfigType.Courses_ShowRecentSearchScreen)
```

4.  Use the hook where you'd like
```tsx
import {useCourseSearchRecentsScreen} from '@frog-pond/app-config`

const MyComponent = () => {
  const shouldShowCourseSearchRecents = useCourseSearchRecentsScreen()
  
  if (shouldShowCourseSearchRecents) {
    // show something to developers
  }
}
```
